tests/e2e/cmp_test.go

```diff
@@ -54,6 +56,33 @@ var _ = Describe("odoCmpE2e", func() {
 			runCmd("odo app create " + appTestName)
 		})
 
+		It("should show an error when ref flag is provided with sources except git", func() {
+			output := runFailCmd(fmt.Sprintf("odo create nodejs cmp-git-%s --local test --ref test", t))
+			Expect(output).To(ContainSubstring("The --ref flag is only valid for --git flag"))
+		})
+
+		It("should create the component from the branch ref when provided", func() {
+			runCmd(fmt.Sprintf("odo create ruby ref-test-%s --git https://github.com/girishramnani/ruby-ex.git --ref develop", t))
+			output := runCmd(fmt.Sprintf("odo url create ref-test-%s", t))
+
+			splitOutput := strings.Split(output, " ")
+			url := strings.TrimSpace(splitOutput[len(splitOutput)-1])
+
+			// the application takes time to come up
+			time.Sleep(10 * time.Second)
```

Comment:
we can't have `time.sleep` in tests... we try to desperately avoid it. Have a look at a few of the `ping` functions that @anmolbabu and @syamgk have created.